cirq/google/engine/engine_job.py

```diff
@@ -365,6 +368,9 @@ def _raise_on_failure(job: quantum.types.QuantumJob) -> None:
     def __iter__(self) -> Iterator[study.TrialResult]:
         return iter(self.results())
 
+    def __getitem__(self, item) -> study.TrialResult:
```

Comment:
Could also use overloads to tell mypy that this acceps `int` or `slice` and returns `study.TrialResult` in the first case but `List[study.TrialResult]` in the latter.

Reply:
yeah should be union of TrialResult and List of TrialResults I think. fixed

Reply:
fixed

Reply:
Union will successfully pass type-checking of this function but for the purposes of type-checking the callers it's somewhat inferior in that it hides from mypy the relationship that exists between argument and return types (namely that `int` gives you a single `study.TrialResult` while a `slice` gives you a sequence).

You can fix this by using [overloads](https://mypy.readthedocs.io/en/stable/more_types.html#function-overloading), probably like:

```python
@overload
def __getitem__(self, item: int) -> study.TrialResult: ...
@overload
def __getitem__(self, item: slice) -> Sequence[study.TrialResult]: ...

def __getitem__(self, item):
    return self.results()[item]
```